src/arithmetic/algebraic_expression/algebraic_expression.c

```diff
@@ -0,0 +1,474 @@
+void AlgebraicExpression_AddChild
+(
+	AlgebraicExpression *root,  // Root to attach child to.
+	AlgebraicExpression *child  // Child node to attach.
+) {
+	assert(root && root->type == AL_OPERATION);
+	root->operation.children = array_append(root->operation.children, child);
+}
+
+// Remove leftmost child node from root.
+AlgebraicExpression *AlgebraicExpression_RemoveLeftmostNode
+(
+	AlgebraicExpression **root  // Root from which to remove left most child.
+) {
+	assert(*root);
+	AlgebraicExpression *prev = *root;
+	AlgebraicExpression *current = *root;
+
+	while(current->type == AL_OPERATION) {
+		prev = current;
+		current = FIRST_CHILD(current);
+	}
+	assert(current->type == AL_OPERAND);
+
+	if(prev == current) {
+		/* Expression is just a single operand
+		 * return operand and update root to NULL. */
+		*root = NULL;
+		return current;
+	}
+
+	/* Removing an operand from an operation
+	 * this might cause a replacement of the operation:
+	 * MUL(A,B) after removing A will become just B
+	 * TRANSPOSE(A) after removing A should become NULL. */
+	if(prev->type == AL_OPERATION) {
+		_AlgebraicExpression_OperationRemoveLeftmostChild(prev);
+		uint child_count = AlgebraicExpression_ChildCount(prev);
+		if(child_count < 2) {
+			if(child_count == 1) {
+				AlgebraicExpression *replacement = _AlgebraicExpression_OperationRemoveRightmostChild(prev);
+				_AlgebraicExpression_InplaceRepurpose(prev, replacement);
+				// Free replacement as it has been copied.
+				rm_free(replacement);
+			} else {
+				assert("for the timebing, we should not be here" && false);
+			}
+		}
+	}
+	return current;
+}
+
+// Remove rightmost child node from root.
+AlgebraicExpression *AlgebraicExpression_RemoveRightmostNode
+(
+	AlgebraicExpression **root  // Root from which to remove left most child.
+) {
+	assert(*root);
+	AlgebraicExpression *prev = *root;
+	AlgebraicExpression *current = *root;
+
+	while(current->type == AL_OPERATION) {
+		prev = current;
+		current = LAST_CHILD(current);
+	}
+	assert(current->type == AL_OPERAND);
+
+	if(prev == current) {
+		/* Expression is just a single operand
+		 * return operand and update root to NULL. */
+		*root = NULL;
+		return current;
+	}
+
+	/* Removing an operand from an operation
+	 * this might cause a replacement of the operation:
+	 * MUL(A,B) after removing A the expression will become just B.
+	 * TRANSPOSE(A) after removing A should become NULL. */
+	if(prev->type == AL_OPERATION) {
+		_AlgebraicExpression_OperationRemoveRightmostChild(prev);
+		uint child_count = AlgebraicExpression_ChildCount(prev);
+		if(child_count < 2) {
+			if(child_count == 1) {
+				AlgebraicExpression *replacement = _AlgebraicExpression_OperationRemoveRightmostChild(prev);
+				_AlgebraicExpression_InplaceRepurpose(prev, replacement);
+				// Free replacement as it has been copied.
+				rm_free(replacement);
+			} else {
+				printf("Warning, operation with no child operands, e.g. empty transpose.\n");
```

Comment:
We get here twice on the test query:
`MATCH (a:actor)-[:act]->(m:movie)<-[:act]-(n:actor{name:'Nicolas Cage'}) RETURN a.name, m.title`
```
#0  AlgebraicExpression_RemoveRightmostNode (root=0x7ffff45400e8)
    at /home/tower-linux/dev/redis/graph/src/arithmetic/algebraic_expression/algebraic_expression.c:372
#1  0x00007ffff59071bd in _AlgebraicExpression_TransposeTranspose (exp=0x7fffedd6fdb0)
    at /home/tower-linux/dev/redis/graph/src/arithmetic/algebraic_expression/algebraic_expression_transpose.c:42
#2  0x00007ffff5907272 in _AlgebraicExpression_TransposeOperation (exp=0x7fffedd6fdb0)
    at /home/tower-linux/dev/redis/graph/src/arithmetic/algebraic_expression/algebraic_expression_transpose.c:69
#3  0x00007ffff590733f in AlgebraicExpression_Transpose (exp=0x7fffedd6fdb0)
    at /home/tower-linux/dev/redis/graph/src/arithmetic/algebraic_expression/algebraic_expression_transpose.c:112
#4  0x00007ffff5907166 in _AlgebraicExpression_TransposeMultiplication (exp=0x7fffedd72e08)
    at /home/tower-linux/dev/redis/graph/src/arithmetic/algebraic_expression/algebraic_expression_transpose.c:31
#5  0x00007ffff5907264 in _AlgebraicExpression_TransposeOperation (exp=0x7fffedd72e08)
    at /home/tower-linux/dev/redis/graph/src/arithmetic/algebraic_expression/algebraic_expression_transpose.c:66
#6  0x00007ffff590733f in AlgebraicExpression_Transpose (exp=0x7fffedd72e08)
    at /home/tower-linux/dev/redis/graph/src/arithmetic/algebraic_expression/algebraic_expression_transpose.c:112
#7  0x00007ffff5960760 in _select_entry_point (qg=0x7ffff6c1f1f0, ae=0x7fffedd72e08, filtered_entities=0x7fffe4a0fda0, bound_vars=0x7fffece0cfc8)
    at /home/tower-linux/dev/redis/graph/src/execution_plan/optimizations/traverse_order.c:280
#8  0x00007ffff5960ada in orderExpressions (qg=0x7ffff6c1f1f0, exps=0x7fffece70b4c, exp_count=2, filters=0x7fffd8201580, bound_vars=0x7fffece0cfc8)
    at /home/tower-linux/dev/redis/graph/src/execution_plan/optimizations/traverse_order.c:364
#9  0x00007ffff592438f in _ExecutionPlan_ProcessQueryGraph (plan=0x7ffff6c15ac0, qg=0x7ffff6c1f1f0, ast=0x7ffff6c7b270, ft=0x7fffd8201580)
    at /home/tower-linux/dev/redis/graph/src/execution_plan/execution_plan.c:30
```

Reply:
That's alright, I've removed the printf,
the only situation in which removing a child from its parent leaving the parent childless is when removing the operand from a transpose operation, e.g. T(T(A)) `_AlgebraicExpression_TransposeTranspose` in which case we're freeing the parent and replacing it with the removed operand.